src/constraints/constraint_node_voltage_angle.jl

```diff
@@ -0,0 +1,109 @@
+#############################################################################
+# Copyright (C) 2017 - 2018  Spine Project
+#
+# This file is part of SpineOpt.
+#
+# SpineOpt is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# SpineOpt is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#############################################################################
+"""
+    constraint_node_voltage_angle(m::Model)
+
+Outer approximation of the non-linear terms.
+"""
+function add_constraint_node_voltage_angle!(m::Model)
+    @fetch node_voltage_angle, connection_flow = m.ext[:variables]
+    t0 = startref(current_window(m))
+    m.ext[:constraints][:node_voltage_angle] = Dict(
+    (connection=conn,node1=n_to,node2=n_from,stochastic_scenario=s,t=t) => @constraint(
+        m,
+        sum(
+        connection_flow[conn,n_from,d_from,s,t]
+        for (conn,n_from,d_from,s,t) in
+            connection_flow_indices(m;
+                connection=conn,
+                node=n_from,
+                direction=direction(:from_node),
+                stochastic_scenario=s,
+                t=t)
+        )
+        -
+        sum(
+        connection_flow[conn,n_to,d_from,s,t]
+        for (conn,n_from,d_from,s,t) in
+            connection_flow_indices(m;
+                connection=conn,
+                node=n_to,
+                direction=direction(:from_node),
+                stochastic_scenario=s,
+                t=t)
+        )
+        ==
+        1/connection_reactance[(connection=conn, stochastic_scenario=s,analysis_time=t0, t=t)] ##reactance or susceptance
+        * connection_reactance_p_u[(connection=conn, stochastic_scenario=s,analysis_time=t0, t=t)]
+        *
+        (sum(
+        node_voltage_angle[n_from,s,t]
+        for (n_from,s,t) in node_voltage_angle_indices(m; node=n_from, stochastic_scenario=s, t=t)
+        )
+        - sum(
+        node_voltage_angle[n_to,s,t]
+        for (n_to,s,t) in node_voltage_angle_indices(m; node=n_to, stochastic_scenario=s, t=t)
+        ))
+        ) for (conn,n_to,n_from,s,t) in constraint_node_voltage_angle_indices(m)
+        )

```

Comment:
It's a bit hard to match opening and closing parentheses here. I suggest using more indentation.